### PR TITLE
[DOC-449] Explain how to use extensions [master]

### PR DESCRIPTION
**Pull request by ashish-macrometa**
- Added documentation for `Extensions` 
- Added example on how to use extensions 
- Fixes [[DOC-449](https://macrometa.atlassian.net/browse/DOC-449)]

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### CODE REVIEW

File: docs/cep/reference/extensions/index.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
+---
+sidebar_position: 3
+title: Extensions
+---
+
+# Extensions
```
Comment (diana-macrometa) [inline docs/cep/reference/extensions/index.md]:
```suggestion
```

Comment (diana-macrometa) [reply docs/cep/reference/extensions/index.md]: Docusaurus automatically turns the title into the H1.

File: docs/cep/reference/extensions/index.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+Extensions enhances Macrometa Stream QL by incorporating additional functionalities in a seamless manner.
```
Comment (diana-macrometa) [inline docs/cep/reference/extensions/index.md]: What is Stream QL? Is it another name for stream workers?

File: docs/cep/reference/extensions/index.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+## Execution extension types
```
Comment (diana-macrometa) [inline docs/cep/reference/extensions/index.md]:
```suggestion
## Execution Extension Types
```

File: docs/cep/reference/extensions/index.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+| Reorder	             | This extension orders out-of-order event arrivals using algorithms such as K-Slack and alpha K-Stack.|
+| Time	                | This extension provides time related functionality such as getting current time, current date, manipulating/formatting dates, etc. |
+| Streaming ML	        | This extension provides streaming machine learning (clustering, classification and regression) on event streams. |
+| Anonymizer	          | The Anonymizer extension provides a function for anonymizing various data types. This function returns a fake value for anonymizing which matches the original data. For example, an email would be replaced with a fake email. |
+| Cache	               | The `cache` extension provides a persistent cache per tenant. |
+| Geo Spatial	         | This extension provides geo data related functionality such as checking whether a given geo coordinate is within a predefined geo-fence, finding distance between 2 geo coordinates etc. |
+| Sentiment	           |This extension performs sentiment analysis using AFINN Wordlist-based approach.|
+
```
Comment (diana-macrometa) [inline docs/cep/reference/extensions/index.md]: Remove all but one blank line between sections.

File: docs/cep/reference/extensions/index.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+
+## IO extensions types
```
Comment (diana-macrometa) [inline docs/cep/reference/extensions/index.md]:
```suggestion
## IO Extensions Types
```

File: docs/cep/reference/extensions/index.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+| ------------- |-------------|
+|Google Pub-Sub | This an extension that receives and publishes events from/to Google Pub/Sub.                                                                                                       |
+|HTTP & HTTPS| The **http extension** extension receives and publishes events via HTTP and HTTPS transports, calls external services, and serves incoming requests and provide synchronous responses. |
+|Apache Kafka (beta)|This an extension that receives and publishes events from/to Apache Kafka.|
+|MQTT|This an extension that receives and publishes events to and from MQTT.|
+|S3|This extension allows to publish events to Amazon AWS S3 buckets.|
+|SSE|This an extension that receives and publishes events from SSE server.|
+
+
+## Format Mappers extension types
```
Comment (diana-macrometa) [inline docs/cep/reference/extensions/index.md]:
```suggestion
## Format Mappers Extension Types
```

File: docs/cep/reference/extensions/index.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+| Format Mappers Extensions     | Description|
+| ------------- |-------------|
+|JSON|This extension converts JSON messages to/from stream processor events.|
+|CSV|This is an extension that converts messages with CSV format to/from stream processor events.|
+|Key-Value|This extension converts events having Key-Value maps to/from stream proceesor events.|
+|Text|This is an extension that converts text messages to/from stream processor events.|
+
+
+#### Example
```
Comment (diana-macrometa) [inline docs/cep/reference/extensions/index.md]:
```suggestion
## Example
```

File: docs/cep/reference/extensions/index.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
+---
+sidebar_position: 3
+title: Extensions
+---
+
+# Extensions
+
+Extensions enhances Macrometa Stream QL by incorporating additional functionalities in a seamless manner.
+
```
Comment (diana-macrometa) [inline docs/cep/reference/extensions/index.md]: This would be a good place to explain to the user why they would want to use extensions, like a few high-level use cases. Something like:

You can use extensions to create stream workers that  (do whatever) and output the results to (format that makes sense).

**Review by diana-macrometa — CHANGES_REQUESTED** [review]
GREAT content. I especially like the example, though I found it a little hard to follow. If you could add a bit more detail for stream worker noobs like me, that would be great.

What I'm still missing is the WIIFM portion. We should explain to users why they might want to use extensions near the beginning as well as how. Remember that we are potentially showing this to a brand new user who might never even have created a stream worker before, so write like you're explaining it to a junior developer from another company.

File: docs/cep/reference/extensions/index.md (hunk 8 above)
Comment (StoyanTc) [reply docs/cep/reference/extensions/index.md]: @diana-macrometa - The `CEP` has core and extensions. The core remains almost invisible for the user except the language and some functions everything else related to `sources` and `sinks` are extensions.

Comment (diana-macrometa) [reply docs/cep/reference/extensions/index.md]: That is helpful. Can you explain to our theoretical noob user how extensions are cool and why they should use them?

Comment (StoyanTc) [reply docs/cep/reference/extensions/index.md]: @diana-macrometa - I'm wandering whether `extension` is the correct term since we neither allow customers to create their own nor get such from e.g. `store`?

Comment (diana-macrometa) [reply docs/cep/reference/extensions/index.md]: I do not know. @durgagokina once said the idea of "extensions" should go away. Whatever we call them, we still need to explain to users how to use these libraries and help them understand their potential. Once you understand what they can do, it's really cool stuff.

File: docs/cep/reference/extensions/index.md (hunk 2 above)
Comment (StoyanTc) [reply docs/cep/reference/extensions/index.md]: @diana-macrometa - 
- `Stream QL`  intentavely refers to `SQL`. This is the language the customers use writing stream workers and it is very close to `SQL`.
- The extensions allow customers to use different sources, sinks (e.g. GDN Streams, Kafka, ActiveMQ) and the appropriate mapping for them (e.g. JSON, XML, CSV) as well as many functions for different kind of processing (e.g. `Math`, `String`, `Statistics`).

File: docs/cep/reference/extensions/index.md (hunk 8 above)
Comment (StoyanTc) [reply docs/cep/reference/extensions/index.md]: @diana-macrometa - Let's keep the `extensions` for now. However I see 3 extensions outside the others:
-[Geospatial](https://macrometa.com/docs/cep/reference/geospatial)
-[Change Data Capture Plugin](https://macrometa.com/docs/cep/change-data-capture-plugin)
-[Store RDBMS Plugin](https://macrometa.com/docs/cep/store-rdbms-plugin)
So, the `plugins` are actually extensions.

File: docs/cep/reference/extensions/index.md (hunk 2 above)
Comment (diana-macrometa) [reply docs/cep/reference/extensions/index.md]: Do we actually talk about Stream QL anywhere? This is the first time I remember seeing it. I'll take a stab at some language based on what you said.

File: docs/cep/reference/extensions/index.md (hunk 8 above)
Comment (diana-macrometa) [reply docs/cep/reference/extensions/index.md]: To make it more fun, they're also inside extensions. Basically, there is a ton of overlap between the `functions` folder and the `extensions` folder, and we need to do an audit and reconcile everything that's there as well as checking to make sure it's all correct.

File: docs/cep/reference/extensions/index.md (hunk 2 above)
Comment (diana-macrometa) [inline docs/cep/reference/extensions/index.md]:
```suggestion
Extensions enhance Macrometa Stream QL, the language used to write stream workers, by seamlessly incorporating additional functionalities.

Extensions allow you to use:

- Different data sources.
- Different sinks, such as GDN Streams, Kafka and ActiveMQ.
- Appropriate format mapping, such as JSON, XML or CSV.
- Different kinds of processing, such as math, string, statistics and geospatial.
```

File: docs/cep/reference/extensions/index.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            "price":"55.6",        
+            "volume":"100"    
+        }
+    }
+
+
+On receiving the event, a query is executed which parses string data types using **Execution extension (math)** into required formats.
+After which the data gets inserted into Kafka `SINK` using **Apache Kafka IO extension**
+
+
```
Comment (diana-macrometa) [inline docs/cep/reference/extensions/index.md]:
```suggestion
```

File: docs/cep/reference/extensions/index.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
```
Comment (diana-macrometa) [inline docs/cep/reference/extensions/index.md]:
```suggestion
```

File: docs/cep/reference/extensions/index.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        "event":{        
+            "symbol":"GDN",        
+            "price":"55.6",        
+            "volume":"100"    
+        }
+    }
+
+
+On receiving the event, a query is executed which parses string data types using **Execution extension (math)** into required formats.
+After which the data gets inserted into Kafka `SINK` using **Apache Kafka IO extension**
```
Comment (diana-macrometa) [inline docs/cep/reference/extensions/index.md]:
```suggestion
After which the data gets inserted into Kafka `SINK` using **Apache Kafka IO extension**.
```

File: docs/cep/reference/extensions/index.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+
+
+```js
+@App:name('TestExtensions') 
+
+CREATE SOURCE FooStream WITH (type='inMemory', topic='stock', map.type='json') (symbol string, price string, volume string);
+
```
Comment (diana-macrometa) [inline docs/cep/reference/extensions/index.md]:
```suggestion
```

File: docs/cep/reference/extensions/index.md (hunk 2 above)
Comment (StoyanTc) [reply docs/cep/reference/extensions/index.md]: Yes, that's new for me as well @diana-macrometa but looks good as interpretation.

**Comment by diana-macrometa** [comment]
@ashish-macrometa, you can click the circling arrows next to a reviewer's name to request a new review.

**Comment by ashish-macrometa** [comment]
moved Sentiment doc to functions folder

File: docs/cep/query-guide/functions/index.md [hunk 13]
```diff
@@ -6,4 +6,143 @@ title: Functions
...
+
+
+Functions enhance Macrometa Stream QL, the language used to write stream workers, by seamlessly incorporating additional capabilities.
+Macrometa Stream QL provides various in-built functions to access and manage event data according to our requirements. Functions can accept 
+zero or more parameters, perform actions and return the result. 
+
+Functions allow you to :
+
+- Use different data sources.
+- Use different sinks, such as GDN Streams, Kafka and ActiveMQ.
```
Comment (diana-macrometa) [inline docs/cep/query-guide/functions/index.md]:
```suggestion
- Use different sinks, such as GDN Streams, Kafka, and ActiveMQ.
```

File: docs/cep/query-guide/functions/index.md [hunk 14]
```diff
@@ -6,4 +6,143 @@ title: Functions
...
+- Appropriate format mapping, such as JSON, XML or CSV.
```
Comment (diana-macrometa) [inline docs/cep/query-guide/functions/index.md]:
```suggestion
- Appropriate format mapping, such as JSON, XML, or CSV.
```

File: docs/cep/query-guide/functions/index.md [hunk 15]
```diff
@@ -6,4 +6,143 @@ title: Functions
...
+- Different kinds of processing, such as math, string, statistics and geospatial.
```
Comment (diana-macrometa) [inline docs/cep/query-guide/functions/index.md]:
```suggestion
- Different kinds of processing, such as math, string, statistics, and geospatial.
```

File: docs/cep/query-guide/functions/index.md [hunk 16]
```diff
@@ -6,4 +6,143 @@ title: Functions
...
+- Data aggregation.
+- Write logical expressions
```
Comment (diana-macrometa) [inline docs/cep/query-guide/functions/index.md]:
```suggestion
- Write logical expressions.
```

File: docs/cep/query-guide/functions/index.md [hunk 17]
```diff
@@ -6,4 +6,143 @@ title: Functions
...
+| Geo Spatial	        | This function provides geo data related functionality such as checking whether a given geo coordinate is within a predefined geo-fence, finding distance between 2 geo coordinates etc. |
+| Sentiment	          |This function performs sentiment analysis using AFINN Wordlist-based approach.|
+
+
+## IO Function Types
+
+| Function     | Description|
+| ------------ |-------------|
+|Google Pub-Sub | This an function that receives and publishes events from/to Google Pub/Sub.                                                                                                      |
+|HTTP & HTTPS| The **http** function receives and publishes events via HTTP and HTTPS transports, calls external services, and serves incoming requests and provide synchronous responses. |
```
Comment (diana-macrometa) [inline docs/cep/query-guide/functions/index.md]:
```suggestion
|HTTP and HTTPS| The **http** function receives and publishes events via HTTP and HTTPS transports, calls external services, and serves incoming requests and provide synchronous responses. |
```

**Review by diana-macrometa — CHANGES_REQUESTED** [review]
Great content! A few minor edits, then we'll be good to go.

**Review by StoyanTc — APPROVED** [review]

**Review by diana-macrometa — APPROVED** [review]
